### PR TITLE
Add sni hostname extension

### LINKED ISSUE #691: Add an `sni_hostname` request extension.
Add a [request extension](https://www.encode.io/httpcore/extensions/#request-extensions) called `"sni_hostname"`, which may be used to control the hostname used when initiating an SSL connection.

This would be an optional override, instead of using the hostname from the URL...

https://github.com/encode/httpcore/blob/5f2e454801f50051d5a66203401a67e5b2a32e9d/httpcore/_async/connection.py#L135-L142

Refs: https://github.com/encode/httpx/discussions/2592#discussioncomment-5859673

### PR DESCRIPTION
**Pull request by karosis88**
closes #691

### CODE REVIEW

**Comment by tomchristie** [comment]
undecided on this, but I'd suggest including the relevant docs update and changelog entry would make sense prior to review.

**Review by tomchristie — COMMENTED** [review]
Nice one, yup. This is a clear win.
Needs adding to the docs and the changelog then it's good to go.

Thanks! 😎

**Review by tomchristie — CHANGES_REQUESTED** [review]

**Review by tomchristie — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Improve logging with per-module logger names. (#690)
 - Resolve race condition during import of `anyio` package. (#692)
 - Enable TCP_NODELAY for all synchronous sockets. (#651)
+- Add `socket_options` argument to `ConnectionPool` and `HTTProxy` classes. (#668)
```
Comment (tomchristie) [inline CHANGELOG.md:12]: Just noticed this isn't in the correct pull request.

**Review by tomchristie — CHANGES_REQUESTED** [review]
One last tweak needed here.

**Comment by karosis88** [comment]
Sorry my bad 🥲

**Review by tomchristie — APPROVED** [review]

**Pull request merged**